File: TheWatch-iOS/Models/GuardReport.swift
import Foundation

// ═══════════════════════════════════════════════════════════════
// Guard Reporting Models
// ═══════════════════════════════════════════════════════════════

enum GuardRole: String, Codable, CaseIterable {
    case professionalGuard = "ProfessionalGuard"
    case neighborhoodWatch = "NeighborhoodWatch"
    case propertyManager = "PropertyManager"
    case campusSecurity = "CampusSecurity"
    case eventStaff = "EventStaff"
    case parkRanger = "ParkRanger"
    case transitSecurity = "TransitSecurity"
    case custom = "Custom"
}

enum ReportCategory: String, Codable, CaseIterable {
    case suspiciousActivity = "SuspiciousActivity"
    case trespass = "Trespass"
    case propertyDamage = "PropertyDamage"
    case disturbance = "Disturbance"
    case medicalObservation = "MedicalObservation"
    case fireHazard = "FireHazard"
    case theftBurglary = "TheftBurglary"
    case assault = "Assault"
    case environmentalHazard = "EnvironmentalHazard"
    case vehicleIncident = "VehicleIncident"
    case welfareConcern = "WelfareConcern"
    case routineObservation = "RoutineObservation"
    case infrastructureIssue = "InfrastructureIssue"
    case other = "Other"
}

enum ReportSeverity: String, Codable, CaseIterable {
    case info = "Info"
    case low = "Low"
    case medium = "Medium"
    case high = "High"
    case critical = "Critical"
}

enum ReportStatus: String, Codable, CaseIterable {
    case draft = "Draft"
    case filed = "Filed"
    case underReview = "UnderReview"
    case escalated = "Escalated"
    case resolved = "Resolved"
    case expired = "Expired"
}

/// A structured report filed by a security guard or watch volunteer.
/// Can be UPGRADED to a full Watch call (SOS dispatch) if the situation warrants.
///
/// Example:
///   let report = try await guardService.fileReport(
///       guardUserId: myUserId,
///       category: .suspiciousActivity,
///       severity: .medium,
///       title: "Person trying doors at loading dock",
///       description: "Male, dark hoodie, tried 3 doors...",
///       latitude: 30.2672, longitude: -97.7431
///   )
///
///   // Situation worsens — escalate:
///   let result = try await guardService.escalateToWatchCall(
///       reportId: report.reportId,
///       reason: "Subject now attempting forced entry"
///   )
///   // result.responseRequestId → responders dispatched
struct GuardReport: Codable, Hashable, Identifiable {
    var id: String { reportId }

    let reportId: String
    let guardUserId: String
    let guardName: String
    let guardRole: GuardRole
    let badgeNumber: String?
    let category: ReportCategory
    let severity: ReportSeverity
    let title: String
    let description: String
    let latitude: Double
    let longitude: Double
    let accuracyMeters: Double?
    let locationDescription: String?
    let status: ReportStatus
    let createdAt: Date
    let filedAt: Date?
    let escalatedAt: Date?
    let resolvedAt: Date?
    let escalatedRequestId: String?
    let escalatedScope: String?
    let evidenceIds: [String]?
    let patrolRouteId: String?
    let postId: String?
    let propertyId: String?
    let reviewedBy: String?
    let reviewNotes: String?
    let resolutionNotes: String?
    let resolvedBy: String?
}

/// Guard enrollment profile.
struct GuardProfile: Codable, Hashable {
    let userId: String
    let name: String
    let role: GuardRole
    let badgeNumber: String?
    let licenseNumber: String?
    let organization: String?
    let canFileReports: Bool
    let canEscalateToWatch: Bool
    let canReviewOtherReports: Bool
    let canAccessCCTV: Bool
    let isOnDuty: Bool
    let shiftStartUtc: Date?
    let shiftEndUtc: Date?
    let enrolledAt: Date
    let lastActiveAt: Date
}

/// Result returned when a guard report is escalated to a Watch call.
struct EscalationResult: Codable, Hashable {
    let reportId: String
    let status: String
    let responseRequestId: String
    let scope: String
    let severity: String
    let escalatedAt: Date?
    let message: String
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/model/GuardReport.kt
package com.thewatch.app.data.model

/**
 * Guard roles for security reporting.
 */
enum class GuardRole {
    ProfessionalGuard,
    NeighborhoodWatch,
    PropertyManager,
    CampusSecurity,
    EventStaff,
    ParkRanger,
    TransitSecurity,
    Custom
}

/**
 * Report categories — determines severity suggestions and supervisor routing.
 */
enum class ReportCategory {
    SuspiciousActivity,
    Trespass,
    PropertyDamage,
    Disturbance,
    MedicalObservation,
    FireHazard,
    TheftBurglary,
    Assault,
    EnvironmentalHazard,
    VehicleIncident,
    WelfareConcern,
    RoutineObservation,
    InfrastructureIssue,
    Other
}

enum class ReportSeverity { Info, Low, Medium, High, Critical }

enum class ReportStatus { Draft, Filed, UnderReview, Escalated, Resolved, Expired }

/**
 * A structured report filed by a security guard.
 * Can be UPGRADED to a full Watch call (SOS dispatch) via the escalate endpoint.
 *
 * Example — file a report then escalate:
 *   val report = guardRepository.fileReport(
 *       guardUserId = myUserId,
 *       category = ReportCategory.SuspiciousActivity,
 *       severity = ReportSeverity.Medium,
 *       title = "Person trying doors at loading dock",
 *       description = "Male, dark hoodie, tried 3 doors...",
 *       latitude = 30.2672,
 *       longitude = -97.7431
 *   )
 *
 *   // Situation worsens — escalate to Watch call
 *   val escalation = guardRepository.escalateToWatchCall(
 *       reportId = report.reportId,
 *       reason = "Subject now attempting forced entry"
 *   )
 *   // escalation.responseRequestId → responders dispatched
 */
data class GuardReport(
    val reportId: String = "",
    val guardUserId: String = "",
    val guardName: String = "",
    val guardRole: String = "NeighborhoodWatch",
    val badgeNumber: String? = null,
    val category: String = "Other",
    val severity: String = "Low",
    val title: String = "",
    val description: String = "",
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
    val accuracyMeters: Double? = null,
    val locationDescription: String? = null,
    val status: String = "Draft",
    val createdAt: String = "",
    val filedAt: String? = null,
    val escalatedAt: String? = null,
    val resolvedAt: String? = null,
    val escalatedRequestId: String? = null,
    val escalatedScope: String? = null,
    val evidenceIds: List<String>? = null,
    val patrolRouteId: String? = null,
    val postId: String? = null,
    val propertyId: String? = null,
    val reviewedBy: String? = null,
    val reviewNotes: String? = null,
    val resolutionNotes: String? = null,
    val resolvedBy: String? = null
)

/**
 * Guard enrollment profile.
 */
data class GuardProfile(
    val userId: String = "",
    val name: String = "",
    val role: String = "NeighborhoodWatch",
    val badgeNumber: String? = null,
    val licenseNumber: String? = null,
    val organization: String? = null,
    val canFileReports: Boolean = true,
    val canEscalateToWatch: Boolean = true,
    val canReviewOtherReports: Boolean = false,
    val canAccessCCTV: Boolean = false,
    val isOnDuty: Boolean = false,
    val shiftStartUtc: String? = null,
    val shiftEndUtc: String? = null
)

/**
 * Result of escalating a report to a Watch call.
 */
data class EscalationResult(
    val reportId: String = "",
    val status: String = "Escalated",
    val responseRequestId: String = "",
    val scope: String = "Neighborhood",
    val severity: String = "High",
    val escalatedAt: String? = null,
    val message: String = ""
)
